src/xdp-utils.c

```diff
@@ -2349,3 +2349,67 @@ xdp_validate_serialized_icon (GVariant  *v,
 
   return TRUE;
 }
+
+GVariant *
+xdp_transform_remote_uris_into_local (GVariant *options)
+{
+  GVariantBuilder out_options;
+  GVariantBuilder uri_list;
+  GVariant *odata = NULL;
+  GVariant *name = NULL;
+  GVariant *uri_value = NULL;
+  GVariant *uris = NULL;
+  GVariant *uri = NULL;
+  gchar *pathtext = NULL;
+  GFile *urifile = NULL;
+  GVfs *gvfs;
+  gchar *uritext = NULL;
+  GString *uristring = NULL;
+  int i, j;
+
+  gvfs = g_vfs_get_default();
+  g_variant_builder_init (&out_options, (const GVariantType *)"a{sv}");
+  g_variant_builder_init (&uri_list, (const GVariantType *)"as");
+  for (i=0; i < g_variant_n_children(options); i ++)
+    {
+      odata = g_variant_get_child_value(options, i);
+      name = g_variant_get_child_value (odata, 0);
+      if (0 != g_strcmp0 ("uris", g_variant_get_string (name, NULL)))
+        {
+          g_variant_builder_add_value (&out_options, odata);
+        }
+      else
+        {
+          uri_value = g_variant_get_child_value (odata, 1);
+          uris = g_variant_get_child_value (uri_value, 0);
+          g_variant_unref (uri_value);
+          for (j=0; j < g_variant_n_children(uris); j++) {
+            uri = g_variant_get_child_value (uris, j);
+            uritext = (gchar *)g_variant_get_string(uri, NULL);
+            if (!g_str_has_prefix (uritext, "file://"))
+              {
+                urifile = g_vfs_get_file_for_uri (gvfs, uritext);
+                pathtext = g_file_get_path (urifile);
+                if (pathtext != NULL)
+                  {
+                    g_variant_unref (uri);
+                    uristring = g_string_new ("");
+                    g_string_printf (uristring, "file://%s", pathtext);
```

Comment:
About g_vfs_* functions, I use them because using GFile methods doesn't work with sftp:// or smb:// URIs (just tested again right now).

About appending "file://", yes, it's quite ugly... fixing it.

Reply:
I removed the `GIO_USE_VFS=local` for this part.

Reply:
Right, so you shouldn't need to call `g_vfs_*` functions, as the gvfs implementation is already the one that should get selected.

Reply:
I was assigned this https://github.com/flatpak/xdg-desktop-portal/issues/213 , and after talking with some co-workers, they suggested implementing this.

Reply:
And they commented about this other problem https://bugzilla.mozilla.org/show_bug.cgi?id=1773624

Reply:
If the app only supports local paths, and can only write to a local path, why is an smb, or ssh URI being passed to xdg-desktop-portal? My guess is that you want the actual file chooser implementation to be more intelligent, because it already knows whether the app wants local paths or not.

In any case, this analysis is missing *a lot* of information. "choosing any of them as the destination folder fails" <- how does it fail? is it trying to write directly to the destination? Is it using GIO or other APIs? Does this happen with say the Firefox Flatpak, and how do you reproduce it to show that it's not a Snap problem?

Reply:
@hadess I attach the test program that I used (with meson/cmake/snapcraft.yaml files) in case you want to check the previous errors.

[test_portal.zip](https://github.com/flatpak/xdg-desktop-portal/files/8965197/test_portal.zip)

Reply:
I already asked for a Flatpak test case, and a number of questions, and you provide me with a Snap testcase that uses a binding (libportal) to a binding (GTK or Qt). You should also try to write reproducers using the barebones APIs, because you're using software that could have bugs on top of other software.

As far as I'm concerned, the problem still isn't root caused.

FWIW, your test case also doesn't convert URIs to paths properly, although that's probably not the problem.

Reply:
@hadess I presumed that the previous message already answered your questions. Anyway, I wrote a little python program that calls the API directly, and did a flatpack with it. The results are the same:

Launching the program from inside the flatpak, I can see the remote mount points:

![imagen](https://user-images.githubusercontent.com/105285003/175311122-0d8dae9b-a97f-4453-9b68-e2a9fb8a9a01.png)

Choosing a local folder in that file selector returns in the test program a translated URI:

![imagen](https://user-images.githubusercontent.com/105285003/175312563-8fcb4af9-7b69-4d49-917c-96876f078f09.png)

but choosing a folder in a remote mount point returns no URIs:

![imagen](https://user-images.githubusercontent.com/105285003/175311675-4f993789-4ac7-47ce-bfd2-7c3b6c8d9083.png)

And also, in the terminal where I have xdg-desktop-portal running, I can see this error:

![imagen](https://user-images.githubusercontent.com/105285003/175311520-e9a1912a-ea60-40da-abbe-64fc18564fab.png)

But now, if I just remove the line `g_setenv ("GIO_USE_VFS", "local", TRUE);` from `xdg-desktop-portal.c` (nothing else, just that line; this patch is not needed), recompile it and launch it again, then, if I try to do the same, there is no error from xdg-desktop-portal, and the python test returns a translated URI, just like in the local case:

![imagen](https://user-images.githubusercontent.com/105285003/175312257-5a94ff85-8bb2-4a76-b39a-032788c29e99.png)

I attach the test program and the json file for creating the flatpak.

[test_portal_python.zip](https://github.com/flatpak/xdg-desktop-portal/files/8967727/test_portal_python.zip)

Reply:
I'm afraid that I don't know enough about the portals to understand why the problem is in the document portal...

Reply:
Watching from the sidelines as one of the original downstream issue reporters, I am wondering why removing `g_setenv ("GIO_USE_VFS", "local", TRUE);` isn't an option even though it fixes the downstream issue?

Reply:
Because we don't long-running daemons that don't deal with files (xdg-desktop-portal should just be passing URIs) to communicate with gvfs. This could also cause xdg-desktop-portal to block on the gvfs daemons availability on startup.

xdg-desktop-portal is a critical component with security implications, and we can't accept patches with a strong root cause analysis. @sergio-costas might want to close this PR and open an issue with the test case instead.

Reply:
Ok.